modules/programs/waybar.nix

```diff
@@ -0,0 +1,388 @@
+      # cfg.settings is a list of Waybar configurations, we need to preserve the index for appropriate warnings
+      allFaultyModules = flip map cfg.settings (settings:
+        let
+          allModules = unique
+            (concatMap (x: attrByPath [ "modules-${x}" ] [ ] settings) [
+              "left"
+              "center"
+              "right"
+            ]);
+          declaredModules = attrNames settings.modules;
+          # Modules declared in `modules` but not referenced in `modules-{left,center,right}`
+          unreferencedModules = subtractLists allModules declaredModules;
+          # Modules listed in modules-{left,center,right} that are not default modules
+          nonDefaultModules = subtractLists defaultModuleNames allModules;
+          # Modules referenced in `modules-{left,center,right}` but not declared in `modules`
+          undefinedModules = subtractLists declaredModules nonDefaultModules;
+          # Check for invalid module names
+          invalidModuleNames =
+            filter (m: !isValidCustomModuleName m) (attrNames settings.modules);
+        in {
+          # The Waybar bar configuration (since config.settings is a list)
+          settings = settings;
+          undef = undefinedModules;
+          unref = unreferencedModules;
+          invalidName = invalidModuleNames;
+        });
+
+      allWarnings = flip concatMap allFaultyModules
+        ({ settings, undef, unref, invalidName }:
+          let
+            unreferenced = map mkUnreferencedModuleWarning unref;
+            undefined = map (mkUndefinedModuleWarning settings) undef;
+            invalid = map mkInvalidModuleNameWarning invalidName;
+          in undefined ++ unreferenced ++ invalid);
+    in allWarnings;
+
+  in mkIf cfg.enable (mkMerge [
+    { home.packages = [ cfg.package ]; }
+    (mkIf (cfg.settings != [ ]) {
+      # Generate warnings about defined but unreferenced modules
+      inherit warnings;
+
+      xdg.configFile."waybar/config".source = configSource;
+    })
+    (mkIf (!isNull cfg.style) {
+      xdg.configFile."waybar/style.css".text = cfg.style;
+    })
+    (mkIf cfg.systemd.enable {
+      systemd.user.services.waybar = {
+        Unit = {
+          Description =
+            "Highly customizable Wayland bar for Sway and Wlroots based compositors.";
+          Documentation = "https://github.com/Alexays/Waybar/wiki";
+          PartOf = [ "graphical-session.target" ];
+          Requisite = [ "dbus.service" ];
+          After = [ "dbus.service" ];
+        };
+
+        Service = {
+          Type = "dbus";
+          BusName = "fr.arouillard.waybar";
+          ExecStart = "${cfg.package}/bin/waybar";
+          Restart = "always";
+          RestartSec = "1sec";
+        };
+
+        Install = {
+          WantedBy = [
+            (if cfg.systemd.withSwayIntegration then
+              "sway-session.target"
```

Comment:
Done. I removed the `withSwayIntegration` option and bind to `graphical-session.target`.
Waybar will automatically exit if not running in a Wayland session.

I'm not sure though what exit code it uses since I don't have an X11 compositor to try it on.
The exit code could be specified in the "allowed" list of exit codes (to prevent the service being restarting multiple times).